File: algorithms/lab07/python/04-bcc/sol1_tarjan.py
#!/usr/bin/env python3
# SPDX-License-Identifier: BSD-3-Clause

import sys


sys.setrecursionlimit(1_000_000)


class Task:
    def __init__(self):
        # n = numar de noduri, m = numar de muchii
        self.n = 0
        self.m = 0

        # adj[i] = lista de adiacenta a nodului i
        self.adj = []

        # ordinea de vizitare
        # found[node] = timpul de start a lui node in parcurgerea DFS
        # in laborator found se numeste idx
        self.found = []

        # low_link[node] = min { found[x] | x este accesibil din node }
        # adica timpul minim al unui nod
        self.low_link = []

        # stiva folosita pentru a reconstrui componentele biconexe
        self.sc: list[tuple[int, int]] = []

        # parent[i] = parintele nodului i
        self.parent = []

        # vector in care retin componentele biconexe
        # all_bccs[i] = componenta biconexa cu indicele i
        self.all_bccs: list[list[int]] = []

    def solve(self) -> None:
        self.read_input()
        self.tarjan()
        self.write_output(self.all_bccs)

    def read_input(self) -> None:
        with open("in", "r", encoding="utf-8") as fin:
            self.n, self.m = map(int, fin.readline().split())
            self.adj = [[] for _ in range(self.n + 1)]
            for _ in range(self.m):
                x, y = map(int, fin.readline().split())
                self.adj[x].append(y)
                self.adj[y].append(x)

    def tarjan(self) -> None:
        self.found = [-1] * (self.n + 1)
        self.low_link = [0] * (self.n + 1)
        self.parent = [0] * (self.n + 1)

        # momentul curent de start
        # pe masura ce vizitam nodurile el va creste (++)
        current_start = 0

        for node in range(1, self.n + 1):
            if self.found[node] == -1:
                # acest nod nu a fost descoperit, deci il putem folosi
                # marcam nodul node ca fiind radacina
                self.parent[node] = 0

                # pornim o noua cautare din nodul respectiv
                current_start = self.dfs(node, current_start)

    def dfs(self, node: int, current_start: int) -> int:
        # incep un nou nod, deci un nou timp de start
        current_start += 1

        # atat found, cat si low_link vor primi valoarea lui current_start
        self.found[node] = current_start
        self.low_link[node] = current_start

        for neigh in self.adj[node]:
            if self.found[neigh] == -1:
                # parintele nodului in care ma duc este chiar nodul curent
                self.parent[neigh] = node

                # retin muchiile de avansare
                self.sc.append((node, neigh))

                # pornesc un nou dfs
                current_start = self.dfs(neigh, current_start)

                # actualizez low_link
                self.low_link[node] = min(self.low_link[node], self.low_link[neigh])

                # daca am gasit un punct de articulatie inseamna ca am descoperit
                # o noua componenta biconexa
                if self.low_link[neigh] >= self.found[node]:
                    self.get_bcc((node, neigh))
            else:
                # !!! graful fiind neorientat trebuie sa ignor muchia catre parinte
                if neigh != self.parent[node]:
                    # am gasit o muchie de intoarcere
                    self.low_link[node] = min(self.low_link[node], self.found[neigh])

        return current_start

    def get_bcc(self, target_edge: tuple[int, int]) -> None:
        # construim o noua componenta biconexa
        current_bcc: list[int] = []

        # extragem muchii din stiva pana am extras muchia target_edge
        while True:
            current_edge = self.sc.pop()

            # adaug capetele muchiei in bcc
            current_bcc.append(current_edge[0])
            current_bcc.append(current_edge[1])

            if current_edge == target_edge:
                break

        # trebuie sa eliminam duplicatele
        current_bcc = sorted(set(current_bcc))

        # current_bcc este o BCC valida
        self.all_bccs.append(current_bcc)

    def write_output(self, result: list[list[int]]) -> None:
        with open("out", "w", encoding="utf-8") as fout:
            fout.write(f"{len(result)}\n")
            for bcc in result:
                fout.write(" ".join(map(str, bcc)) + " \n")


def main() -> None:
    task = Task()
    task.solve()


if __name__ == "__main__":
    main()

File: algorithms/lab07/rust/04-bcc/sol1_tarjan.rs
// SPDX-License-Identifier: BSD-3-Clause

use std::cmp::min;
use std::fs;

#[derive(Default)]
struct Task {
	// n = numar de noduri, m = numar de muchii
	n: usize,

	// adj[i] = lista de adiacenta a nodului i
	adj: Vec<Vec<usize>>,

	// ordinea de vizitare
	// found[node] = timpul de start a lui node in parcurgerea DFS
	// in laborator found se numeste idx
	found: Vec<i32>,

	// low_link[node] = min { found[x] | x este accesibil din node }
	// adica timpul minim al unui nod
	low_link: Vec<i32>,

	// parent[i] = parintele nodului i
	parent: Vec<usize>,

	// stiva folosita pentru a reconstrui componentele biconexe
	stack_edges: Vec<(usize, usize)>,

	// vector in care retin componentele biconexe
	// all_bccs[i] = componenta biconexa cu indicele i
	all_bccs: Vec<Vec<usize>>,
}

impl Task {
	fn solve(&mut self) {
		self.read_input();
		self.tarjan();
		self.write_output();
	}

	fn read_input(&mut self) {
		let data = fs::read_to_string("in").expect("failed to read input");
		let mut it = data.split_whitespace();

		self.n = it.next().unwrap().parse::<usize>().unwrap();
		let m = it.next().unwrap().parse::<usize>().unwrap();
		self.adj = vec![Vec::new(); self.n + 1];

		for _ in 0..m {
			let x = it.next().unwrap().parse::<usize>().unwrap();
			let y = it.next().unwrap().parse::<usize>().unwrap();
			self.adj[x].push(y);
			self.adj[y].push(x);
		}
	}

	fn tarjan(&mut self) {
		// PASUL 1: initializez rezultatele
		self.found = vec![-1; self.n + 1];
		self.low_link = vec![0; self.n + 1];
		self.parent = vec![0; self.n + 1];
		self.stack_edges.clear();
		self.all_bccs.clear();

		// momentul curent de start
		// pe masura ce vizitam nodurile el va creste (++)
		let mut current_start = 0;
		// PASUL 2: vizitez toate nodurile
		for node in 1..=self.n {
			if self.found[node] == -1 {
				// acest nod nu a fost descoperit, deci il putem folosi
				// marcam nodul node ca fiind radacina
				self.parent[node] = 0;

				// PASUL 3: pornesc o noua parcurgere DFS pe acest subarbore
				self.dfs(node, &mut current_start);
			}
		}
	}

	fn dfs(&mut self, node: usize, current_start: &mut i32) {
		// PASUL 1: un nod nou este vizitat - incrementez timestamp-ul
		*current_start += 1;
		self.found[node] = *current_start;
		self.low_link[node] = *current_start;

		// PASUL 2: vizitez fiecare vecin
		let neighbors = self.adj[node].clone();
		for neigh in neighbors {
			// PASUL 3: verific daca neigh este deja vizitat
			if self.found[neigh] == -1 {
				// PASUL 4: salvez parintele
				self.parent[neigh] = node;

				// retin muchiile de avansare
				self.stack_edges.push((node, neigh));

				// PASUL 5: vizitez recursiv subarborele copilului
				self.dfs(neigh, current_start);

				// PASUL 6: actualizez low_link[node] cu informatiile obtinute prin neigh
				self.low_link[node] = min(self.low_link[node], self.low_link[neigh]);

				// daca am gasit un punct de articulatie inseamna ca am descoperit
				// o noua componenta biconexa
				if self.low_link[neigh] >= self.found[node] {
					self.get_bcc((node, neigh));
				}
			} else if neigh != self.parent[node] {
				// am gasit o muchie de intoarcere
				self.low_link[node] = min(self.low_link[node], self.found[neigh]);
			}
		}
	}

	fn get_bcc(&mut self, target_edge: (usize, usize)) {
		// construim o noua componenta biconexa
		let mut current_bcc = Vec::new();

		// extragem muchii din stiva pana am extras muchia target_edge
		loop {
			let edge = self.stack_edges.pop().unwrap();

			// adaug capetele muchiei in bcc
			current_bcc.push(edge.0);
			current_bcc.push(edge.1);

			if edge == target_edge {
				break;
			}
		}

		// trebuie sa eliminam duplicatele
		current_bcc.sort_unstable();
		current_bcc.dedup();

		// current_bcc este o BCC valida
		self.all_bccs.push(current_bcc);
	}

	fn write_output(&self) {
		let mut out = String::new();
		out.push_str(&format!("{}\n", self.all_bccs.len()));
		for bcc in &self.all_bccs {
			for &node in bcc {
				out.push_str(&format!("{} ", node));
			}
			out.push('\n');
		}
		fs::write("out", out).expect("failed to write output");
	}
}

fn main() {
	let mut task = Box::new(Task::default());
	task.solve();
}
